anvil/src/udev.rs

```diff
@@ -700,6 +719,62 @@ impl AnvilState<UdevData> {
         }
     }
 
+    fn vblank(&mut self, dev_id: DrmNode, crtc: crtc::Handle, metadata: &mut Option<DrmEventMetadata>) {
+        let device_backend = match self.backend_data.backends.get_mut(&dev_id) {
+            Some(backend) => backend,
+            None => {
+                error!(self.log, "Trying to render on non-existent backend {}", dev_id);
+                return;
+            }
+        };
+        let surfaces = device_backend.surfaces.borrow();
+        let surface = match surfaces.get(&crtc) {
+            Some(surface) => surface,
+            None => {
+                error!(self.log, "Trying to render on non-existent backend {}", dev_id);
+                return;
+            }
+        };
+
+        let output = match self
+            .space
+            .outputs()
+            .find(|o| {
+                o.user_data().get::<UdevOutputId>()
+                    == Some(&UdevOutputId {
+                        device_id: surface.borrow().device_id,
+                        crtc,
+                    })
+            })
+            .cloned()
+        {
+            Some(output) => output,
+            None => return,
+        };
+
+        let report = match surface.borrow_mut().output_render_report.take() {
+            Some(report) => report,
+            None => return,
+        };
+        std::mem::drop(surfaces);
+
+        let tp = metadata.as_ref().and_then(|metadata| match metadata.time {
+            smithay::backend::drm::DrmEventTime::Monotonic(tp) => Some(tp),
+            smithay::backend::drm::DrmEventTime::Realtime(_) => None,
+        });
+        let seq = metadata.as_ref().map(|metadata| metadata.sequence).unwrap_or(0);
+
+        let (tv_sec, tv_nsec) = if let Some(tp) = tp {
+            let zero = unsafe { std::mem::zeroed::<std::time::Instant>() };
```

Comment:
I have used this hack before, but this needs a SAFETY note, that explains, that this relies on the internal representation of `std::time::Instant` on unix.

Reply:
yeah, right. I definitely do not like that at all and atm the only thing we do is to convert the `Duration` in the source to an `Instant` and then back to a `Duration` here, so maybe we should remove the hack in the source and find a different solution?

Reply:
Yeah, we should just fix this to return a `Duration` instead of doing this hack twice: https://github.com/cmeissl/smithay/blob/feature/wp_presentation/src/backend/drm/device/mod.rs#L390